File: TwoPointers/ValidPalindrome/vp.go
package main

import "fmt"

func isPalindrome(s string) bool {
	l, r := 0, len(s)-1
	for l < r {
		if !isAlphaNum(s[l]) {
			l++
			continue
		}
		if !isAlphaNum(s[r]) {
			r--
			continue
		}
		if lower(s[l]) != lower(s[r]) {
			return false
		}
		l++
		r--
	}
	return true
}

func isAlphaNum(c byte) bool {
	b := 'A' <= c && c <= 'Z'
	b2 := 'a' <= c && c <= 'z'
	b3 := '0' <= c && c <= '9'
	return b || b2 || b3
}

func lower(c byte) byte {
	var bit int
	if 'A' <= c && c <= 'Z' {
		bit = 1
	}
	mask := byte(bit << 5)
	return c | mask
}

func main() {
	s := "Was it a car or a cat I saw?"
	fmt.Printf("T1: %t\n", isPalindrome(s))

	s = "tab a cat"
	fmt.Printf("T2: %t\n", isPalindrome(s))

	s = "race car"
	fmt.Printf("T3: %t\n", isPalindrome(s))
}

File: TwoPointers/ValidPalindrome/vp.py
class Solution:
    # NOTE: time = O(n), space = O(1)
    def isPalindrome(self, s: str) -> bool:
        l = 0
        r = len(s) - 1
        while l < r:
            if not s[l].isalnum():
                l += 1
                continue
            if not s[r].isalnum():
                r -= 1
                continue
            if s[l].lower() != s[r].lower():
                return False
            l += 1
            r -= 1

        return True

    # NOTE: time = O(n), space = O(n)
    # def isPalindrome(self, s: str) -> bool:
    #     s = "".join(c for c in s if c.isalnum()).lower()
    #     return s == s[::-1]


if __name__ == "__main__":
    print("MAIN:")
    sol = Solution()

    s = "Was it a car or a cat I saw?"
    r = sol.isPalindrome(s)
    print(f"T1: {r}")

    s = "tab a cat"
    r = sol.isPalindrome(s)
    print(f"T2: {r}")

    s = "race car"
    r = sol.isPalindrome(s)
    print(f"T3: {r}")
